File: VB/Spreadsheet_Document_Api_Part_3/CodeExamples/CustomXmlPartActions.vb
Imports DevExpress.Spreadsheet
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Xml

Namespace SpreadsheetControl_API_Part03.CodeExamples
	Friend Class CustomXmlPartActions
		Private Shared Sub StoreCustomXmlPart(ByVal workbook As IWorkbook)
			' #Region "#StoreCustomXmlPart"
			workbook.Worksheets(0).Cells("A1").Value = "Custom Xml Test"

			' Add an empty custom XML part.
			Dim part As ICustomXmlPart = workbook.CustomXmlParts.Add()
			Dim elem As XmlElement = part.CustomXmlPartDocument.CreateElement("Person")
			elem.InnerText = "Stephen Edwards"
			part.CustomXmlPartDocument.AppendChild(elem)

			' Add an XML part created from string.
			Dim xmlString As String = "<?xml version=""1.0"" encoding=""UTF-8""?>" & ControlChars.CrLf &
					"<whitepaper>" & ControlChars.CrLf &
					  "<contact>" & ControlChars.CrLf &
						"<firstname>Roger</firstname>" & ControlChars.CrLf &
						"<lastname>Edwards</lastname>" & ControlChars.CrLf &
						"<phone>832-433-0025</phone>" & ControlChars.CrLf &
						"<address>1657 Wines Lane Houston, TX 77099</address>" & ControlChars.CrLf &
					  "</contact>" & ControlChars.CrLf &
					  "<date>2016-05-18</date>" & ControlChars.CrLf &
					"</whitepaper>"
			workbook.CustomXmlParts.Add(xmlString)

			' Add an XML part loaded from a file.
			Dim xmlDoc As New XmlDocument()
			xmlDoc.Load("Documents\fishes.xml")
			workbook.CustomXmlParts.Add(xmlDoc)
			workbook.SaveDocument("Documents\CustomXmlTest.xlsx")
			System.IO.File.Copy("Documents\CustomXmlTest.xlsx", "Documents\CustomXmlTest.xlsx.zip", True)
			System.Diagnostics.Process.Start("Documents\CustomXmlTest.xlsx.zip")
			' #End Region ' #StoreCustomXmlPart
		End Sub

		Private Shared Sub ObtainCustomXmlPart(ByVal workbook As IWorkbook)
			' #Region "#ObtainCustomXmlPart"
			' Load a document from a file.
			workbook.LoadDocument("Documents\CustomXml.xlsx")

			' Access a custom XML file stored in the document.
			Dim xmlDoc As XmlDocument = workbook.CustomXmlParts(0).CustomXmlPartDocument

			' Retrieve a hyperlink from the XML file And display it in the document.
			Dim nsmgr As New XmlNamespaceManager(xmlDoc.NameTable)
			Dim xPathString As String = "//Fish[Category='Cod']/ScientificClassification/Reference"
			Dim xmlNode As XmlNode = xmlDoc.DocumentElement.SelectSingleNode(xPathString, nsmgr)
			Dim hLink As String = xmlNode.InnerText
			workbook.Worksheets(0).Hyperlinks.Add(workbook.Worksheets(0).Cells("A2"), hLink, True)
			' #End Region ' #ObtainCustomXmlPart
		End Sub

		Private Shared Sub ModifyCustomXmlPart(ByVal workbook As IWorkbook)
			' #Region "#ModifyCustomXmlPart"
			' Load a document from a file.
			workbook.LoadDocument("Documents\CustomXml.xlsx")

			' Access a custom XML file stored in the document.
			Dim xmlDoc As XmlDocument = workbook.CustomXmlParts(1).CustomXmlPartDocument

			' Retrieve XML nodes that match a specific expression.
			Dim nsmgr As New XmlNamespaceManager(xmlDoc.NameTable)
			Dim xPathString As String = "//whitepaper/contact[firstname='Roger']/firstname"
			Dim xmlNodes As XmlNodeList = xmlDoc.DocumentElement.SelectNodes(xPathString, nsmgr)

			' Modify XML nodes and display them in the document.
			For Each node As XmlNode In xmlNodes
				node.InnerText = "Stephen"
			Next node
			workbook.SaveDocument("Documents\CustomXmlRogerStephen.xlsx")
			workbook.Worksheets(0).Cells("A2").Value = xmlDoc.FirstChild.FirstChild.FirstChild.InnerText
			' #End Region ' #ModifyCustomXmlPart
		End Sub
	End Class
End Namespace

File: CS/Spreadsheet_Document_Api_Part_3/CodeExamples/CustomXmlPartActions.cs
﻿using DevExpress.Spreadsheet;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace SpreadsheetControl_API_Part03.CodeExamples
{
    class CustomXmlPartActions
    {
        static void StoreCustomXmlPart(IWorkbook workbook)
        {
            #region #StoreCustomXmlPart
            workbook.Worksheets[0].Cells["A1"].Value = "Custom Xml Test";

            // Add an empty custom XML part.
            ICustomXmlPart part = workbook.CustomXmlParts.Add();
            XmlElement elem = part.CustomXmlPartDocument.CreateElement("Person");
            elem.InnerText = "Stephen Edwards";
            part.CustomXmlPartDocument.AppendChild(elem);

            // Add an XML part created from string.
            string xmlString = @"<?xml version=""1.0"" encoding=""UTF-8""?>
                                    <whitepaper>
                                       <contact>
                                          <firstname>Roger</firstname>
                                          <lastname>Edwards</lastname>
                                          <phone>832-433-0025</phone>
                                          <address>1657 Wines Lane Houston, TX 77099</address>
                                       </contact>
                                       <date>2016-05-18</date>
                                    </whitepaper>";
            workbook.CustomXmlParts.Add(xmlString);

            // Add an XML part loaded from a file.
            XmlDocument xmlDoc = new XmlDocument();
            xmlDoc.Load("Documents\\fishes.xml");
            workbook.CustomXmlParts.Add(xmlDoc);
            workbook.SaveDocument("Documents\\CustomXmlTest.xlsx");
            System.IO.File.Copy("Documents\\CustomXmlTest.xlsx", "Documents\\CustomXmlTest.xlsx.zip", true);
            System.Diagnostics.Process.Start("Documents\\CustomXmlTest.xlsx.zip");
            #endregion #StoreCustomXmlPart
        }

        static void ObtainCustomXmlPart(IWorkbook workbook)
        {
            #region #ObtainCustomXmlPart
            // Load a document from a file.
            workbook.LoadDocument("Documents\\CustomXml.xlsx");

            // Access a custom XML file stored in the document.
            XmlDocument xmlDoc = workbook.CustomXmlParts[0].CustomXmlPartDocument;

            // Retrieve a hyperlink from the XML file and display it in the document.
            XmlNamespaceManager nsmgr = new XmlNamespaceManager(xmlDoc.NameTable);
            string xPathString = "//Fish[Category='Cod']/ScientificClassification/Reference";
            XmlNode xmlNode = xmlDoc.DocumentElement.SelectSingleNode(xPathString, nsmgr);
            string hLink = xmlNode.InnerText;
            workbook.Worksheets[0].Hyperlinks.Add(workbook.Worksheets[0].Cells["A2"], hLink, true);
            #endregion #ObtainCustomXmlPart
        }

        static void ModifyCustomXmlPart(IWorkbook workbook)
        {
            #region #ModifyCustomXmlPart
            // Load a document from a file.
            workbook.LoadDocument("Documents\\CustomXml.xlsx");

            // Access a custom XML file stored in the document.
            XmlDocument xmlDoc = workbook.CustomXmlParts[1].CustomXmlPartDocument;
            
            // Retrieve XML nodes that match a specific expression.
            XmlNamespaceManager nsmgr = new XmlNamespaceManager(xmlDoc.NameTable);
            string xPathString = "//whitepaper/contact[firstname='Roger']/firstname";
            XmlNodeList xmlNodes = xmlDoc.DocumentElement.SelectNodes(xPathString, nsmgr);
            
            // Modify XML nodes and display them in the document.
            foreach (XmlNode node in xmlNodes) node.InnerText = "Stephen";
            workbook.SaveDocument("Documents\\CustomXmlRogerStephen.xlsx");
            workbook.Worksheets[0].Cells["A2"].Value = xmlDoc.FirstChild.FirstChild.FirstChild.InnerText;
            #endregion #ModifyCustomXmlPart
        }
    }
}
